File: backend/app/schemas/leaderboard.py
"""Leaderboard schemas."""
from uuid import UUID
from datetime import datetime
from typing import Optional
import enum
from pydantic import BaseModel, Field

from app.models.training_session import TrainingMode


class LeaderboardPeriod(str, enum.Enum):
    """Leaderboard period enum."""
    DAY = "day"
    WEEK = "week"
    MONTH = "month"
    ALL = "all"


class LeaderboardEntry(BaseModel):
    """Leaderboard entry schema."""
    user_id: UUID
    username: str
    wpm: int
    cpm: int
    accuracy: float
    errors: int
    correct_chars: int
    duration_sec: int
    created_at: datetime

    class Config:
        from_attributes = True


class LeaderboardResponse(BaseModel):
    """Leaderboard response schema."""
    items: list[LeaderboardEntry]
    total: int
    mode: TrainingMode
    period: str

File: frontend/src/api/leaderboard.ts
import apiClient from './client'

export type LeaderboardPeriod = 'day' | 'week' | 'month' | 'all'
export type TrainingMode = 'text' | 'code'

export interface LeaderboardEntry {
  user_id: string
  username: string
  wpm: number
  cpm: number
  accuracy: number
  errors: number
  correct_chars: number
  duration_sec: number
  created_at: string
}

export interface LeaderboardResponse {
  items: LeaderboardEntry[]
  total: number
  mode: TrainingMode
  period: string
}

export const leaderboardApi = {
  get: async (params: {
    mode: TrainingMode
    period?: LeaderboardPeriod
    sort_by?: 'wpm' | 'accuracy'
    limit?: number
    offset?: number
  }): Promise<LeaderboardResponse> => {
    const response = await apiClient.get<LeaderboardResponse>('/api/leaderboard', {
      params,
    })
    return response.data
  },
}

